File: ios/Sources/Ratakierros/Models/Track.swift
import Foundation

struct Track: Codable, Identifiable, Hashable {
    let id: Int64
    let lipasId: Int64?
    let name: String?
    let lat: Double
    let lon: Double
    let city: String?
    let suburb: String?
    let address: String?
    let postalCode: String?
    let surface: String?
    let trackLengthM: Int64?
    let lanes: Int64?
    let status: String?
    let typeCode: Int64?
    let distanceM: Double?
    let record: Double?

    enum CodingKeys: String, CodingKey {
        case id
        case lipasId = "lipas_id"
        case name
        case lat
        case lon
        case city
        case suburb
        case address
        case postalCode = "postal_code"
        case surface
        case trackLengthM = "track_length_m"
        case lanes
        case status
        case typeCode = "type_code"
        case distanceM = "distance_m"
        case record
    }

    var displayName: String { name ?? "—" }

    var distanceLabel: String? {
        guard let d = distanceM else { return nil }
        if d < 1000 { return String(format: "%.0f m", d) }
        return String(format: "%.1f km", d / 1000.0)
    }

    var recordLabel: String? {
        guard let r = record else { return nil }
        return formatSeconds(r)
    }
}

func formatSeconds(_ seconds: Double) -> String {
    let total = max(0, seconds)
    let minutes = Int(total) / 60
    let remainder = total - Double(minutes * 60)
    return String(format: "%d:%05.2f", minutes, remainder)
}

File: android/app/src/main/kotlin/fi/ratakierros/model/Track.kt
package fi.ratakierros.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class Track(
    val id: Long,
    @SerialName("lipas_id") val lipasId: Long? = null,
    val name: String? = null,
    val lat: Double,
    val lon: Double,
    val city: String? = null,
    val suburb: String? = null,
    val address: String? = null,
    @SerialName("postal_code") val postalCode: String? = null,
    val surface: String? = null,
    @SerialName("track_length_m") val trackLengthM: Long? = null,
    val lanes: Long? = null,
    val status: String? = null,
    @SerialName("type_code") val typeCode: Long? = null,
    @SerialName("distance_m") val distanceM: Double? = null,
    val record: Double? = null,
) {
    val displayName: String get() = name ?: "—"

    val distanceLabel: String? get() = distanceM?.let {
        if (it < 1000) String.format("%.0f m", it) else String.format("%.1f km", it / 1000.0)
    }

    val recordLabel: String? get() = record?.let { formatSeconds(it) }
}

fun formatSeconds(seconds: Double): String {
    val total = if (seconds < 0) 0.0 else seconds
    val minutes = total.toInt() / 60
    val remainder = total - minutes * 60
    return String.format("%d:%05.2f", minutes, remainder)
}
